packages/react-async/src/reducer.js

```diff
@@ -53,6 +54,17 @@ export const reducer = (state, { type, payload, meta }) => {
         finishedAt: new Date(),
         ...getStatusProps(statusTypes.rejected),
       }
+    case actionTypes.reinitialize:
+      return {
+        ...state,
+        data: undefined,
+        value: undefined,
+        error: undefined,
+        startedAt: undefined,
+        finishedAt: undefined,
+        ...getStatusProps(statusTypes.initial),
+        counter: meta.counter,
+      }
```

Comment:
This should use the `init` function above, which is what specifies the initial reducer state. That's what I'd expect it to reinitialize to. I think you'll need to pass props/options as the `payload` so you can just do:
```js
case actionTypes.reinitialize:
  return init(payload)
```

Reply:
I agree with you. But I got a question: `status` will be `pending`, not `initial` if the `init` is fired with `promiseFn`, is it correct behavior?

Reply:
We should take care to honor the props. I think this should mean:
- If `initialValue` is provided (usually server-side) we should use that and set `startedAt` to undefined and keep `finishedAt` as-is. We don't want to use `new Date()` in this case.
- If `promise` is provided (which cannot be restarted), `startedAt` should stay as-is (since we'll be reusing the original promise, not restarting it) and `finishedAt` should be undefined (it will be resolved/rejected immediately after anyway).
- If `promiseFn` is provided, `startedAt` can be `new Date()` because we will restart it immediately after, so `init()` does the right thing there.
- `getInitialStatus` already does the right thing.

Hence I came up with the following:

```js
case actionTypes.reinitialize:
  const newState = init(payload)
  if (payload.initialValue) newState.finishedAt = state.finishedAt
  if (payload.promise) newState.startedAt = state.startedAt
  return newState
```

Reply:
Cool, I agree with you mostly. But there is one thing confusing me still.
> getInitialStatus already does the right thing.

`getInitialStatus` will return `pending` status if `promise` is provided because of [this line](
https://github.com/ghengeveld/react-async/blob/master/packages/react-async/src/status.js#L11).

Accordingly, `isInitial` will not be `true` and `isPending` will be `true`.

Is this correct behavior?

Reply:
I expected that when the `reinitialize` is invoked, the status should be `initial`:
```
isInitial => true
isPending => false
isFulfilled => false
isRejected => false
```
But you think it should be `pending`. 
```
isInitial => false
isPending => true
isFulfilled => false
isRejected => false
```

Is this correct?

Reply:
That sounds good. I noticed you changed the initial status with `promiseFn` to be `initial` instead of `pending`, which is okay. On mount the promiseFn will be invoked, causing it to change to `pending` immediately after. I think that is strictly more correct.

Reply:
In my mind, the purpose of `reinitialize` is not to change the state to `initial`, but to "reset" the component as if it was destroyed and recreated. That also means the `promiseFn` should be invoked again as it is when first mounted. Right now `reinitialize` only resets the state, but doesn't trigger the reload.

Reply:
I see :)
I expected to use `reinitialize` and then `reload` or `run`.
Otherwise, I would use just `reload` mostly.

Reply:
I will wrap up this PR as the following

> Reset the component as if it was destroyed and recreated

- reinitialized with promiseFn -> pending
- reinitialized with promise -> pending
- reinitialized with initialValue -> fulfilled

Reply:
Okay, fair enough. At least we have this for future reference. I'm impressed you went this far only to conclude it's better for React Async to stay simple. Not a lot of people think about it that way, or would be willing to sidestep their time investment. Thanks a lot for contributing, I really appreciate it.

PS If you're looking for a new remote challenge perhaps [Chroma](https://hichroma.com/) is something for you.

Reply:
That's nice to hear, thanks! Chromatic looks very interesting.

I made a small custom hook, as a simple reference, which was I wanted.
- Storybook: https://noelyoo.github.io/use-async-fn/?path=/story/asyncfn--promisefn
- Github: https://github.com/noelyoo/use-async-fn

I think this can be useful for someone in some context.